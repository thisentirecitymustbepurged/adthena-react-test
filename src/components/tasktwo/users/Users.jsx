import React, { useMemo } from 'react';
import User from './user';
import './Users.scss';

const groupBy = (arr, key) => arr.reduce((acc, el) => {
  if (acc[el[key]]) {
    acc[el[key]].push(el);
  } else {
    acc[el[key]] = [el];
  }

  return acc;
}, {});

const mapTodos = (users, todosByUserId) => users.map(
  user => ({ ...user, todos: todosByUserId[user.id] })
);

const Users = ({ users, todos }) => {
  const todosByUserId = useMemo(() => groupBy(todos, 'userId'), [todos]);
  const usersWithTodos = useMemo(() => mapTodos(users, todosByUserId), [todosByUserId, users]);

  return (
    <div className="users">
      <div className="users__header">
        <span>Click on a user to toggle visibility of todos</span>
      </div>
      {usersWithTodos.map(user => <User {...{ ...user, key: user.id }} />)}
    </div>
  );
};

export default Users;
