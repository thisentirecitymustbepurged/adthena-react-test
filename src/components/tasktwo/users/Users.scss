.users {
  &__header {
    > * {
      margin-bottom: 15px;
    }
  }
  &__filter {
    > * {
      margin-right: 10px;
    }
    input {
      padding: 7px;
      border: 1px solid rgba(0,0,0,0.5);
    }
  }
}