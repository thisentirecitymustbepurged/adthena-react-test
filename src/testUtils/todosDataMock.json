[
  {
    "userId": 1,
    "id": 1,
    "title": "delectus aut autem",
    "completed": false
  },
  {
    "userId": 1,
    "id": 2,
    "title": "et officia qui",
    "completed": false
  },
  {
    "userId": 2,
    "id": 28,
    "title": "nesciunt totam sit blanditiis sit",
    "completed": false
  },
  {
    "userId": 2,
    "id": 29,
    "title": "laborum aut in quam",
    "completed": false
  },
  {
    "userId": 2,
    "id": 30,
    "title": "nemo perspiciatis repellat ut dolor libero commodi blanditiis omnis",
    "completed": true
  }
]
