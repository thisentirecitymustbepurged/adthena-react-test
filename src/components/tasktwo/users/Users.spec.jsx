import React from "react";
import { shallow } from "enzyme";
import Users from "@components/tasktwo/users/Users";
import usersDataMock from "@testUtils/usersDataMock";
import todosDataMock from "@testUtils/todosDataMock";

const render = props => shallow(<Users {...props} />);

describe("User component", () => {
  it("should match snapshot", () => {
    const wrapper = render({
      users: usersDataMock,
      todos: todosDataMock
    });

    expect(wrapper).toMatchSnapshot();
  });

  it("should render users", () => {
    const wrapper = render({
      users: usersDataMock,
      todos: todosDataMock
    });

    const users = wrapper.find("User");

    expect(users.length).toBe(2);
  });

  it("should serialize and pass todos to the User component", () => {
    const wrapper = render({
      users: usersDataMock,
      todos: todosDataMock
    });

    const firstUser = wrapper.find("User").first();

    expect(firstUser.props().todos).toEqual([
      { completed: false, id: 1, title: "delectus aut autem", userId: 1 },
      { completed: false, id: 2, title: "et officia qui", userId: 1}
    ]);
  });
});
