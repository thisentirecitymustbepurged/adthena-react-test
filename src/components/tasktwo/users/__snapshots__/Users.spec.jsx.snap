// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component should match snapshot 1`] = `
<div
  className="users"
>
  <div
    className="users__header"
  >
    <span>
      Click on a user to toggle visibility of todos
    </span>
  </div>
  <User
    address={
      Object {
        "city": "Gwenborough",
        "geo": Object {
          "lat": "-37.3159",
          "lng": "81.1496",
        },
        "street": "Kulas Light",
        "suite": "Apt. 556",
        "zipcode": "92998-3874",
      }
    }
    company={
      Object {
        "bs": "harness real-time e-markets",
        "catchPhrase": "Multi-layered client-server neural-net",
        "name": "Romaguera-Crona",
      }
    }
    email="Sincere@april.biz"
    id={1}
    key="1"
    name="Leanne Graham"
    phone="1-770-736-8031 x56442"
    todos={
      Array [
        Object {
          "completed": false,
          "id": 1,
          "title": "delectus aut autem",
          "userId": 1,
        },
        Object {
          "completed": false,
          "id": 2,
          "title": "et officia qui",
          "userId": 1,
        },
      ]
    }
    username="Bret"
    website="hildegard.org"
  />
  <User
    address={
      Object {
        "city": "Wisokyburgh",
        "geo": Object {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      }
    }
    company={
      Object {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      }
    }
    email="Shanna@melissa.tv"
    id={2}
    key="2"
    name="Ervin Howell"
    phone="010-692-6593 x09125"
    todos={
      Array [
        Object {
          "completed": false,
          "id": 28,
          "title": "nesciunt totam sit blanditiis sit",
          "userId": 2,
        },
        Object {
          "completed": false,
          "id": 29,
          "title": "laborum aut in quam",
          "userId": 2,
        },
        Object {
          "completed": true,
          "id": 30,
          "title": "nemo perspiciatis repellat ut dolor libero commodi blanditiis omnis",
          "userId": 2,
        },
      ]
    }
    username="Antonette"
    website="anastasia.net"
  />
</div>
`;
